drivers/modbus/modbus_rtu.c

```diff
@@ -0,0 +1,688 @@
+#include "thread.h"
+#include "timex.h"
+#include "ztimer.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#ifdef MODULE_MODBUS_RTU
+
+enum {
+    ID = 0,     /* ID field */
+    FUNC,       /* Function code position */
+    ADD_HI,     /* Address high byte */
+    ADD_LO,     /* Address low byte */
+    NB_HI,      /* Number of coils or registers high byte */
+    NB_LO,      /* Number of coils or registers low byte */
+    BYTE_CNT    /* byte counter */
+};
+
+static uint16_t crc_calculate(const uint8_t *buf, size_t len)
+{
+    return ucrc16_calc_le(buf, len, 0xA001, 0xFFFF);
+}
+
+static bool crc_check(const uint8_t *buf, size_t len)
+{
+    assert(len > 2);
+
+    uint16_t actual = crc_calculate(buf, len - 2);
+    uint16_t expected = (buf[len - 2] << 0) |
+                        (buf[len - 1] << 8);
+
+    if (actual != expected) {
+        DEBUG("[modbus_rtu] crc_check: actual 0x%04x != expected 0x%04x\n", actual, expected);
+    }
+
+    return actual == expected;
+}
+
+static inline uint16_t read_address(modbus_rtu_t *modbus)
+{
+    return byteorder_bebuftohs(modbus->buffer + ADD_HI);
+}
+
+static inline void write_address(modbus_rtu_t *modbus, uint16_t addr)
+{
+    byteorder_htobebufs(modbus->buffer + ADD_HI, addr);
+}
+
+static inline uint16_t read_count(modbus_rtu_t *modbus)
+{
+    return byteorder_bebuftohs(modbus->buffer + NB_HI);
+}
+
+static inline void write_count(modbus_rtu_t *modbus, uint16_t count)
+{
+    byteorder_htobebufs(modbus->buffer + NB_HI, count);
+}
+
+static inline void write_size(modbus_rtu_t *modbus, uint8_t size, bool request)
+{
+    modbus->buffer[request ? BYTE_CNT : 2] = size;
+}
+
+static inline uint8_t read_size(modbus_rtu_t *modbus, bool request)
+{
+    return modbus->buffer[request ? BYTE_CNT : 2];
+}
+
+static inline void enable_rts(modbus_rtu_t *modbus)
+{
+    if (gpio_is_valid(modbus->params->pin_rts)) {
+        gpio_write(modbus->params->pin_rts, modbus->params->pin_rts_enable);
+    }
+}
+
+static inline void disable_rts(modbus_rtu_t *modbus)
+{
+    if (gpio_is_valid(modbus->params->pin_rts)) {
+        gpio_write(modbus->params->pin_rts, !modbus->params->pin_rts_enable);
+    }
+}
+
+static inline void send(modbus_rtu_t *modbus)
+{
+    uint16_t crc = crc_calculate(modbus->buffer, modbus->buffer_size);
+
+    modbus->buffer[modbus->buffer_size++] = (crc >> 0) & 0xff;
+    modbus->buffer[modbus->buffer_size++] = (crc >> 8) & 0xff;
+
+    enable_rts(modbus);
+    DEBUG("[modbus_rtu] send: %d bytes\n", modbus->buffer_size);
+    uart_write(modbus->params->uart, modbus->buffer, modbus->buffer_size);
+    disable_rts(modbus);
+}
+
+static int receive(modbus_rtu_t *modbus, uint8_t expected_size)
+{
+    msg_t msg;
+
+    while (modbus->buffer_size < expected_size) {
+        if (ztimer_msg_receive_timeout(ZTIMER_USEC, &msg, modbus->byte_timeout) < 0) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static void uart_cb(void *arg, uint8_t data)
+{
+    static msg_t msg;
```

Comment:
This is never written to. Is this just a dummy message to wake-up the thread?

Reply:
yes, it is

it's not uncommon in riot to do so, but other synchronization mechanisms might be a better choice